### ISSUE TITLE
wxWebSessionCURL crashes on `Close()`

### ISSUE THREAD

**Issue by ryancog** [wxWidgets/wxWidgets#24969]
### Description

Whenever using wxWebSession on Linux w/ wxGTK (and thus the wxWebSessionCURL backend), the app crashes whenever the session's `Close()` function is called, although running requests should have completed by that point.

### Expected behavior

The program cleans up web session resources and continues as normal, instead the program consistently crashes with a segfault.

### Stack trace:

```
#0  0x00000000005a1ef6 in SocketPoller::StartPolling(int, int) ()
#1  0x00000000005a974e in wxWebSessionCURL::ProcessSocketCallback(void*, int, int) ()
#2  0x00000000005a9886 in wxWebSessionCURL::SocketCallback(void*, int, int, void*, void*) ()
#3  0x00007ffff6a2aa3e in ??? () at /lib64/libcurl.so.4
#4  0x00007ffff69f3c63 in ??? () at /lib64/libcurl.so.4
#5  0x00007ffff69f3dfa in ??? () at /lib64/libcurl.so.4
#6  0x00007ffff69f4f12 in ??? () at /lib64/libcurl.so.4
#7  0x00007ffff6a2a739 in curl_multi_cleanup () at /lib64/libcurl.so.4
#8  0x00000000005a8cb3 in wxWebSessionCURL::~wxWebSessionCURL() ()
#9  0x00000000005a8e1d in non-virtual thunk to wxWebSessionCURL::~wxWebSessionCURL() ()
#10 0x00000000005987b9 in wxWebSessionBase::Close() ()
#11 0x000000000047440e in WebTest::OnInit (this=0x9c5a80)
    at [redacted]/webtest/main.cpp:39
#12 0x00000000004cf513 in wxEntry(int&, wchar_t**) ()
#13 0x00000000004744e0 in main (argc=1, argv=0x7fffffffd748)
    at [redacted]/webtest/main.cpp:45

```

### Example program:

```
#include <thread>
#include <wx/app.h>
#include <wx/webrequest.h>

class WebTest : public wxApp {
public:
    bool OnInit() override {
        wxEvtHandler handler;

        bool requestComplete{false};
        auto handleRequestEvent{[&](wxWebRequestEvent& evt) {
            if (evt.GetState() == wxWebRequest::State_Completed) {
                // Temp file copied to working path for use later
            }

            switch (evt.GetState()) {
                case wxWebRequestBase::State_Unauthorized:
                case wxWebRequestBase::State_Completed:
                case wxWebRequestBase::State_Failed:
                case wxWebRequestBase::State_Cancelled:
                    requestComplete = true;
                default: break;
            }
        }};

        const wxString pullFrom{"https://google.com/"};
        auto webSession{wxWebSession::New()};
        auto request{webSession.CreateRequest(&handler, pullFrom)};
        request.SetStorage(wxWebRequest::Storage_File);
        handler.Bind(wxEVT_WEBREQUEST_STATE, handleRequestEvent);
        request.Start();

        while(not requestComplete) {
            wxYield();
            std::this_thread::sleep_for(std::chrono::milliseconds(50));
        }

        handler.Unbind(wxEVT_WEBREQUEST_STATE, handleRequestEvent);
        webSession.Close();

        return false;
    }
};

wxIMPLEMENT_APP(WebTest);
```

### Platform and version information

- wxWidgets version you use: master (tested on 5edd516a)
- wxWidgets port you use: wxGTK
- OS and its version: OpenSUSE Tumbleweed
  + GTK version: 3.25.43
  + Which GDK is used: Wayland
  + Desktop environment: GNOME

**Comment by NewPagodi** [comment]
I think the problematic lines are in webrequest_curl.cpp

```
wxWebSessionCURL::~wxWebSessionCURL()
{
    delete m_socketPoller;

    if ( m_handle )
        curl_multi_cleanup(m_handle);
}
```
curl_multi_cleanup must be calling the socketPoller to flush any remaining data.  I'm not sure why this hasn't come up before.  Did curl recently change the behavior for curl_multi_cleanup?

 I think we can just change the order to:
```
wxWebSessionCURL::~wxWebSessionCURL()
{
    if ( m_handle )
        curl_multi_cleanup(m_handle);

    delete m_socketPoller;
}
```

Does that sound right?

**Comment by ryancog** [comment]
That logically makes sense to me, though I'm far from an expert in using curl. I can make those changes locally and see if that resolves it. :)

**Comment by ryancog** [comment]
Well, in the minimal example above, that change seems to make things work.

However in my actual program, which closely mirrors the example (though not directly in `OnInit()`), when the end of scope is reached (according to gdb at least), the destructor of the request is called, and that seems to result in `StopActiveTransfer()` being called, which eventually segfaults (or SIGFPE, though that seems to be irrelevant since it's coming from `m_activeTransfers::find()`, so an implementation detail of `std::unordered_map<>`)

Best as I can tell, that being a member of the `Close()`'d `wxWebSessionCURL` is the issue. What the call path looks like I'm not sure, I've not stared at it for long enough to decipher it all, but here's the relevant bit of the backtrace:

```
#0  0x00000000005d5697 in std::_Hashtable<void*, std::pair<void* const, wxWebRequestCURL*>, std::allocator<std::pair<void* const, wxWebRequestCURL*> >, std::__detail::_Select1st, std::equal_to<void*>, std::hash<void*>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<false, false, true> >::find(void* const&) [clone .isra.0] ()
#1  0x00000000005dcb8a in wxWebSessionCURL::StopActiveTransfer(void*) ()
#2  0x00000000005dccf1 in wxWebRequestCURL::~wxWebRequestCURL() ()
#3  0x00000000005dce29 in wxWebRequestCURL::~wxWebRequestCURL() ()
#4  0x0000000000478dca in wxWebRequest::~wxWebRequest (this=0x7fffffffc1c0, this=<optimized out>)
    at [redacted]/include/wx-3.3/wx/webrequest.h:260
```

This seems problematic to me because it would indicate the request object must be destroyed before the session can be, which feels counterintuitive. As far as the explanation of the API goes, it sounds to me like the expected behavior is that once the request is complete, the session can be `Close()`'d.

Now, wrapping the request object inside an unnamed scope, and putting the `webSession.Close()` call outside of that, things seem to work as expected.

In my opinion, it seems like if it's expected that the request needs to be destroyed before the session can be closed, then there should be an explicit way to do that, beyond scoping the request, that way it's more obvious. (A note in the documentation also wouldn't hurt I imagine, if this is indeed expected)

If this is not expected, then I suppose the question becomes how to deal with it, and seems to start to get into the weeds of the wxWidgets implementation.

**Comment by ryancog** [comment]
For added clarity, this is a snippet from the working application code (with scope added):
```
    auto webSession{wxWebSession::New()};
    {
        auto request{webSession.CreateRequest(getEventHandler(), pullFrom)};
        request.SetStorage(wxWebRequest::Storage_File);
        getEventHandler()->Bind(wxEVT_WEBREQUEST_STATE, handleRequestEvent);
        request.Start();

        logger.info("Pulling version data...");
        while(not requestComplete) {
            if (not prog->Pulse("Pulling version data...")) {
                logger.info("Canceled.");
                return nullopt;
            }

            wxYield();
            std::this_thread::sleep_for(std::chrono::milliseconds(50));
        }

        getEventHandler()->Unbind(wxEVT_WEBREQUEST_STATE, handleRequestEvent);

        if (request.GetState() != wxWebRequestSync::State_Completed) {
            logger.info("Data pull failed.");
            return nullopt;
        }
    }
    webSession.Close();
```

**Comment by vadz** [comment]
One thing I can definitely say is that `wxWebSession::Close()` is not necessary, i.e. that it will be done by the dtor anyhow.

I _believe_ that `wxWebRequest` must not survive the session it is associated with, but I think that a much better way to ensure that this happens would be to increase the ref count of `wxWebSessionImplPtr` in `wxWebRequestImpl`, which should be relative simple to do: we'd just need to change its `m_sessionImpl` type to `wxWebSessionImplPtr`. We'd have to check that this doesn't result in any reference loops, however.

Finally, to return to the reported crash, I think it must be a consequence of 4e0fca8ab9 (Fix crash when connection is refused in wxWebRequestCURL, 2024-10-27) somehow, even though I don't see how did it change this, but it's the only change which was done in this code recently and a bug as bad as this one must be relatively new. In any case, I agree that the lifetimes are wrong here and that we need to keep `m_socketPoller` alive as long as libcurl can still use it.

**Comment by vadz** [comment]
What is your libcurl version? I'm surprised I can't reproduce this with my libcurl 7.88.1: there are no calls to `m_socketPoller` from `curl_multi_cleanup()` if there are no active requests.

We still should keep it alive, just in case (or at least reset to null and check for this) but I don't understand why does the behaviour differ.

**Comment by vadz** [comment]
I've created #24970 because I've convinced myself that this can do no harm but I'd still like to understand why I don't see anything wrong happening (even when running under valgrind) with the initial example, please let me know if you have any ideas.

And, of course, please test the PR if you can. TIA!

(and automatic ref counting will have to wait, as this would be more work as it would have to be done for all backends, not just this one — it wouldn't make sense if the code using it worked but the same code crashed under Windows or macOS)

**Comment by ryancog** [comment]
> One thing I can definitely say is that `wxWebSession::Close()` is not necessary, i.e. that it will be done by the dtor anyhow.

Excellent, thanks for clarifying :)

> Finally, to return to the reported crash, I think it must be a consequence of https://github.com/wxWidgets/wxWidgets/commit/4e0fca8ab9756989598d07b41e672af86eac7092 (Fix crash when connection is refused in wxWebRequestCURL, 2024-10-27) somehow, even though I don't see how did it change this, but it's the only change which was done in this code recently and a bug as bad as this one must be relatively new. In any case, I agree that the lifetimes are wrong here and that we need to keep m_socketPoller alive as long as libcurl can still use it.

So, fun fact, I originally had a version of master that was from before this commit. The crash was experienced on that version, and in searching for a solution I found that. I thought perhaps it was related (and it couldn't hurt to try), so I pulled recent changes, including that one, and the crash was still there.

All that is to say that I experienced the issue using a master version prior to that commit, and so I don't believe that's where it started.

libcurl's behavior changing for some reason (I'm lucky I suppose) seems more plausible to me, since as you said this is a pretty major bug for having been unnoticed thus far.

> What is your libcurl version? I'm surprised I can't reproduce this with my libcurl 7.88.1

My version is libcurl 8.10.1... the difference in major version would be suspect there, perhaps the behavior did change? 

> there are no calls to m_socketPoller from curl_multi_cleanup() if there are no active requests.

The second crash I experienced, stemming from `StopActiveTransfer` would indicate that the issue is perhaps because for some reason the request is still active in some capacity? Perhaps `StopActiveTransfer` is called unconditionally, but in any case it seems like an odd coincidence.

I can test the PR here in a bit, I appreciate that little addition to the doc btw.

**Comment by ryancog** [comment]
@vadz Your PR compiles successfully for me and displays (perhaps unsurprisingly) the same results as with my application of NewPagodi's fix.

The minimal example program still runs without issue, and my program crashes at the same point within `StopActiveTransfer`. So at least things are consistent.

**Comment by ryancog** [comment]
Upon further inspection, I'm not sure why the backtrace doesn't reflect this, but it seems that the call looks something like:
1. `~wxWebRequestCURL()`
1. `m_sessionCURL->RequestHasTerminated(this)`
1. `StopActiveTransfer(curl)`

At which point the `m_sessionCURL` is no longer valid so everything from there on out is undefined behavior, and the hashmap `find()` is where it happens to crash for me (because it seems this is the first thing of substance to happen).

As I'm typing this I'm realizing this is probably not new information and you've realized this already.

So then, the suggestion of making that `m_sessionCURL` object reference counted makes sense to me, but I do wonder if it wouldn't make sense to just make it more obvious that the requests do in fact need to end their life before the session can end.

It seems like it would perhaps be surprising for me if there were any potential "leftover" traces from the session after it supposedly disappeared in my application code, as the reference counting could potentially cause.

I would prefer to maybe have a way to, if I couldn't simply let the request go out of scope as I now know it can, either have the request clean itself up on completion (or failure?) so that the session's `Close()` could be called, or perhaps add a `Close()`-type method to `wxWebRequest` to explicitly clean itself up, before presumably calling session `Close()`.

This way, it's more clear to the application programmer what is happening and the lifetimes of these objects and how they interact is more obvious?

That's just my suggestion from my perspective. Either solution makes sense to me, but it just seems more intuitive to be more explicit, requiring the application programmer to be aware of how the lifetimes work and to have better control over them.

**Comment by vadz** [comment]
FWIW I've confirmed that the crash happens, without the fix in the associated PR, when using libcurl 8.10 and later (8.11.0 too) even though it doesn't happen with 7.88.1 in Debian Bookworm. I could probably run git-bisect on libcurl to find the exact commit which changed this, but this doesn't really seem to be worth it.

**Comment by vadz** [comment]
I've added the code to keep the session alive for as long as necessary now, with this fix the original example works too, so this will be closed when the PR is merged.

**Comment by ryancog** [comment]
Added the premature close to my application and can confirm the fix works for me as well.
